unic/char/range/src/range.rs

```diff
@@ -0,0 +1,203 @@
+use std::char;
+use std::collections::Bound;
+use std::ops::Range;
+use step;
+
+const SURROGATE_RANGE: Range<u32> = 0xD800..0xE000;
+
+/// A range of unicode code points.
+///
+/// The members of this struct are public for const initialization by `chars!(..=)` only.
+/// They should be considered unstable private API that may change at any time.
+/// If you decide to use them anyway, make sure to note the safety notes.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct CharRange {
+    /// The lowest uniterated character (inclusive).
+    ///
+    /// Iteration is finished if this is higher than `high`.
+    ///
+    /// # Safety
+    ///
+    /// This is not guaranteed to always be a valid character. Check before using!
+    /// Note that `high` _is_ guaranteed to be a valid character,
+    /// so this will always be a valid character when iteration is not yet finished.
+    #[doc(hidden)]
+    pub low: char,
+
+    /// The highest uniterated character (inclusive).
+    ///
+    /// Iteration is finished if this is lower than `low`.
+    #[doc(hidden)]
+    pub high: char,
+}
+
+/// Constructors
+impl CharRange {
+    /// Construct a closed range of characters.
+    pub fn closed(start: char, stop: char) -> CharRange {
+        CharRange {
+            low: start,
+            high: stop,
+        }
+    }
+
+    /// Construct a half open (right) range of characters.
+    pub fn open_right(start: char, stop: char) -> CharRange {
+        let mut range = CharRange::closed(start, stop);
+        range.step_backward();
+        range
+    }
+
+    /// Construct a half open (left) range of characters.
+    pub fn open_left(start: char, stop: char) -> CharRange {
+        let mut range = CharRange::closed(start, stop);
+        range.step_forward();
+        range
+    }
+
+    /// Construct a fully open range of characters.
+    pub fn open(start: char, stop: char) -> CharRange {
+        let mut range = CharRange::closed(start, stop);
+        range.step_forward();
+        range.step_backward();
+        range
+    }
+
+    /// Construct a range of characters from bounds.
+    pub fn bound(mut start: Bound<char>, mut stop: Bound<char>) -> CharRange {
```

Comment:
Of note: you can pass an imutable owned value here, because it's taken by value.

But this formulation makes more sense than what I have. 👍